### PR TITLE
PLUGIN-595/GroupByAggregator append functions with conditional functions

### PR DESCRIPTION
**Pull request by ardian4**
Append functions that have ability to define condition to check if the record meets the criteria

JiraTicket: https://cdap.atlassian.net/browse/PLUGIN-595

### CODE REVIEW

File: core-plugins/docs/GroupByAggregate-batchaggregator.md [hunk 1]
```diff
@@ -24,13 +27,19 @@ output records will have a ``user`` field and a ``numActions`` field. (Macro-ena
...
 be called. Aggregates are specified using the syntax `name:function(field)[, other aggregates]`.
-For example, ``avgPrice:avg(price),cheapest:min(price)`` will calculate two aggregates.
+For example, ``avgPrice:avg(price),cheapest:min(price),countPricesHigherThan:countIf(price):condition(price>500)``
+will calculate three aggregates.
 The first will create a field called ``avgPrice`` that is the average of all ``price`` fields in the group.
 The second will create a field called ``cheapest`` that contains the minimum ``price`` field in the group.
+The third will create a field ``countPricesHigherThan`` that contains the number of all ``price`` fields in the group 
+that meet the condition bigger than 500.
 The count function differs from count(*) in that it contains non-null values of a specific field,
 while count(*) will count all records regardless of value. (Macro-enabled)
```
Comment (flakrimjusufi) [inline core-plugins/docs/GroupByAggregate-batchaggregator.md:45]: Please add more examples with newly added aggregate functions.

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByAggregator.java [hunk 2]
```diff
@@ -175,9 +197,50 @@ public void prepareRun(BatchAggregatorContext context) throws Exception {
...
+   */
+  private boolean fieldExistsInOutputSchema(List<String> path, Schema schema) {
+    if (path.size() == 0) {
+      return false;
+    }
+    Schema.Field currentField = schema.getField(path.get(0));
+    if (currentField == null) {
+      return false;
+    }
+    if (path.size() > 1) {
```
Comment (flakrimjusufi) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByAggregator.java:254]: Can you provide an example where size of path will be > 1 ?

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByConfig.java [hunk 3]
```diff
@@ -147,13 +168,26 @@ public GroupByConfig() {
           "Could not find closing ')' in function '%s'. Functions must be specified as function(field).",
           functionAndField));
       }
-      String field = functionAndField.substring(leftParanIdx + 1, functionAndField.length() - 1).trim();
+      int conditionIndex = functionAndField.toLowerCase().indexOf("condition(");
```
Comment (flakrimjusufi) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByConfig.java:171]: Are we sure that `"condition("` will always be the same and won't change?

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByConfig.java [hunk 4]
```diff
@@ -147,13 +168,26 @@ public GroupByConfig() {
...
           "Invalid function '%s'. A field must be given as an argument.", functionAndField));
       }
-
-      functionInfos.add(new FunctionInfo(name, field, function));
+      if (conditionIndex == -1 && function.isConditional()) {
+        throw new IllegalArgumentException("Missing 'condition' property for conditional function.");
+      }
+      String functionCondition = null;
+      if (conditionIndex != -1) {
+        functionCondition = functionAndField.substring(conditionIndex + 10, functionAndField.length() - 1);
```
Comment (flakrimjusufi) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByConfig.java:188]: Can we add a comment before this line explaining why we need to perform a substring function, especially why we need to do something like this: 

`functionCondition = functionAndField.substring(conditionIndex + 10, functionAndField.length() - 1);`

Please also include an example of how we are expecting `functionCondition` to look like before and after substring.

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/GroupByTestRun.java [hunk 5]
```diff
@@ -469,4 +469,48 @@ private void ingestData(String datasetName, Schema schema) throws Exception {
...
+
+  @Test
+  public void testGroupByConditionValidationFailure() {
+    Schema purchaseSchema = Schema.recordOf(
+      "purchase",
+      Schema.Field.of("ts", Schema.of(Schema.Type.LONG)),
+      Schema.Field.of("user", Schema.of(Schema.Type.STRING)),
+      Schema.Field.of("item", Schema.of(Schema.Type.STRING)),
+      Schema.Field.of("price", Schema.of(Schema.Type.DOUBLE)));
+    Schema aggSchema = Schema.recordOf(
```
Comment (flakrimjusufi) [inline core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/GroupByTestRun.java]: aggSchema is not used, we should remove it.

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/GroupByTestRun.java [hunk 6]
```diff
@@ -469,4 +469,48 @@ private void ingestData(String datasetName, Schema schema) throws Exception {
...
+      "purchase.agg",
+      Schema.Field.of("item", Schema.of(Schema.Type.STRING)),
+      Schema.Field.of("maxPrice", Schema.arrayOf(Schema.of(Schema.Type.STRING))));
+
+    GroupByConfig groupByConfig = new GroupByConfig("item",  "maxPrice:MaxIf(price):condition(test<5)");
+    GroupByAggregator groupByAggregator = new GroupByAggregator(groupByConfig);
+    MockPipelineConfigurer mockConfigurer = new MockPipelineConfigurer(purchaseSchema, Collections.emptyMap());
+    try {
+      groupByAggregator.configurePipeline(mockConfigurer);
+      Assert.fail("Condition with wrong input field that does not exists in schema should fail validation");
```
Comment (flakrimjusufi) [inline core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/GroupByTestRun.java:507]: Your test is not catching the scenario you are trying to test and it's not failing for the reason it should fail. 

I think you are trying to catch `validationException` here, not just `Exception`.

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/GroupByTestRun.java [hunk 7]
```diff
@@ -469,4 +469,48 @@ private void ingestData(String datasetName, Schema schema) throws Exception {
...
+    } catch (Exception e) {
+      // expected
```
Comment (flakrimjusufi) [inline core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/GroupByTestRun.java]: In assertion we should also catch the expected error message and failure size.

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/AvgTest.java [hunk 8]
```diff
@@ -1,12 +1,12 @@
 /*
  * Copyright © 2016-2019 Cask Data, Inc.
- *  
+ *
```
Comment (flakrimjusufi) [inline core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/AvgTest.java:3]: Let's remove those unnecessary changes from this file.

**Review by flakrimjusufi — CHANGES_REQUESTED** [review]

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByAggregator.java (hunk 2 above)
Comment (ardian4) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByAggregator.java:254]: Updated, example is in method javadoc.

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByConfig.java (hunk 3 above)
Comment (ardian4) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByConfig.java:171]: Yes

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIf.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    this.condition = condition;
+  }
+
+  @Override
+  public void mergeValue(StructuredRecord record) {
+    if (!meetCondition(record, condition)) {
+      return;
+    }
+    super.mergeValue(record);
+  }
```
Comment (rmstar) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIf.java:42]: do you also need to override [mergeAggregates](https://github.com/cdapio/hydrator-plugins/blob/ac2ae149e30c28a936f0860c633bba03ada634d8/core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/LogicalAnd.java#L52)?

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/JexlCondition.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+    if (result instanceof Boolean) {
+      return (boolean) result;
+    } else {
+      throw new IllegalArgumentException("incorrect condition");
+    }
+  }
+
+  /**
+   * Return value of a field from a record based on provided bath
```
Comment (rmstar) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/JexlCondition.java]: bath -> path

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/JexlCondition.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  /**
+   * Generates JexlScript containing condition
+   *
+   * @return {@link JexlScript}
+   */
+  private JexlScript getScript() {
+    if (engine == null) {
+      engine = new JexlBuilder().cache(1024).strict(true).silent(false).create();
```
Comment (rmstar) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/JexlCondition.java:99]: Can this be called concurrently from multiple threads? If yes, you would need some synchronization here.

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/JexlCondition.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  /**
+   * This method extracts variables from a given condition
+   *
+   * @param condition string representation of Jexl condition
+   * @return set of lists representing full path of each variable
+   */
+  public static Set<List<String>> getVariables(String condition) {
+    JexlEngine engine = new JexlBuilder().cache(1024).strict(true).silent(false).create();
```
Comment (rmstar) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/JexlCondition.java:124]: Can't we use this.engine, this.script here?

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/AvgIfTest.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  @Test
+  public void testLongCondition() {
+    String condition = "x<101L";
+
+    Schema schema = Schema.recordOf("test", Schema.Field.of("x", Schema.of(Schema.Type.LONG)));
+    Avg avg = new AvgIf("x", Schema.of(Schema.Type.LONG), JexlCondition.of(condition));
+    Avg avg1 = new AvgIf("x", Schema.of(Schema.Type.LONG), JexlCondition.of(condition));
+    testFunction(avg, schema, avg1, 0d, 0);
```
Comment (rmstar) [inline core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/AvgIfTest.java:46]: Add test case where the condition evalues to false. Similar comment for other tests.

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/AvgTest.java (hunk 8 above)
Comment (rmstar) [reply core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/AvgTest.java:3]: Please fix

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/CountIfTest.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+package io.cdap.plugin.batch.aggregator.function;
+
+import io.cdap.cdap.api.data.schema.Schema;
+import org.junit.Test;
+
+import java.util.Arrays;
+
+/**
+ *
```
Comment (rmstar) [inline core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/CountIfTest.java]: remove empty comment

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIfTest.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+import io.cdap.cdap.api.data.schema.Schema;
+import org.junit.Test;
+
+import java.util.Arrays;
+
+public class LogicalAndIfTest extends LogicalAndTest {
+
+  @Test
+  public void testCondition() {
+    String condition = "x>1";
```
Comment (rmstar) [inline core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIfTest.java]: what does x>1 mean when x is a boolean?

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIf.java (hunk 9 above)
Comment (ardian4) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIf.java:42]: Condition acts more like a filter for records, since the records are filtered out during mergeValue and in the mergeAggregates we only have the results of Aggregates

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/JexlCondition.java (hunk 12 above)
Comment (ardian4) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/JexlCondition.java:124]: this method is static and we use during validation, so we can not use this.engine or this.script

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/AvgIfTest.java (hunk 13 above)
Comment (ardian4) [reply core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/AvgIfTest.java:46]: Tests are updated

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIfTest.java (hunk 15 above)
Comment (ardian4) [reply core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIfTest.java]: In this case JEXL will evaluate to false, but we have changed the test to be more understandable

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/JexlCondition.java (hunk 11 above)
Comment (ardian4) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/JexlCondition.java:99]: I have updated the code

File: core-plugins/docs/GroupByAggregate-batchaggregator.md [hunk 16]
```diff
@@ -40,16 +49,17 @@ framework will decide on the number to use.
...
+into consideration prices lower than 0.50.
 
 ```json
     {
         "name": "GroupByAggregate",
         "type": "batchaggregator",
         "properties": {
             "groupByFields": "user,item",
-            "aggregates": "totalSpent:sum(price),numPurchased:count(*)"
+            "aggregates": "totalSpent:sum(price),numPurchased:count(*),avgItemPrice:avgIf(price):condition(price>0.50)"
```
Comment (rmstar) [inline core-plugins/docs/GroupByAggregate-batchaggregator.md]: should be `condition(price>=0.50)` because `third one calculates the average price of each item not taking 
into consideration prices lower than 0.50`. Also, please include some prices < 0.50 in your example dataset.

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIf.java (hunk 9 above)
Comment (rmstar) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIf.java:42]: What happens when a LogicalAndIf aggregate is merged with a LogicalAnd aggregate that contains elements that don't match the condition. Is such a scenario possible?
e.g:  `LogicalAndIf("x", JexlCondition.of("x == false))` is merged with `LogicalAndIf("x")`

Similar question for other aggregate types.

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByAggregator.java [hunk 17]
```diff
@@ -140,6 +161,7 @@ public void validate(Schema inputSchema, List<String> groupByFields,
         validateCountDistinct(inputField, collector, collectorFieldName);
       }
     }
+    validateConditionalFunctions(inputSchema, conf.getAggregates(), collector);
```
Comment (rmstar) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByAggregator.java:164]: Can you please verify if this works as expected in a pipeline with macros? In particular, please test this when the previous stage's output schema is a macro

Comment (ardian4) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByAggregator.java:164]: We tested when previous stage's output schema is macro and it seems ok

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIf.java (hunk 9 above)
Comment (ardian4) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIf.java:42]: We think this scenario is not possible because aggregates are grouped by their aliases and we can not have multiple aggregates with the same alias. So is not possible to merge two different functions. 

Below we can see that aggregates are grouped by their aliases.
https://github.com/cdapio/hydrator-plugins/blob/44065a6fa6121a86ea7bb5c7c2de3c3f10c7d28f/core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByAggregator.java#L371-L378

Also below we can see that they are merged based by key which we can see from above that key is the alias.
https://github.com/cdapio/hydrator-plugins/blob/44065a6fa6121a86ea7bb5c7c2de3c3f10c7d28f/core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByAggregator.java#L343-L345

Comment (rmstar) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/function/LogicalAndIf.java:42]: I see, thanks!

**Review by rmstar — APPROVED** [review]
Please rebase on top of latest develop branch and squash commits

**Comment by rmstar** [comment]
Test failures look unrelated to this PR
```
2021-05-04T05:12:27.0102009Z testHBaseSource(io.cdap.plugin.HBaseTest)  Time elapsed: 3.438 sec  <<< ERROR!
2021-05-04T05:12:27.0104065Z java.lang.NoSuchMethodError: org.apache.hadoop.fs.FSOutputSummer.<init>(Ljava/util/zip/Checksum;II)V
2021-05-04T05:12:27.0105964Z 	at org.apache.hadoop.hdfs.DFSOutputStream.<init>(DFSOutputStream.java:1342)
2021-05-04T05:12:27.0107572Z 	at org.apache.hadoop.hdfs.DFSOutputStream.<init>(DFSOutputStream.java:1371)
2021-05-04T05:12:27.0109047Z 	at org.apache.hadoop.hdfs.DFSOutputStream.newStreamForCreate(DFSOutputStream.java:1403)
2021-05-04T05:12:27.0110433Z 	at org.apache.hadoop.hdfs.DFSClient.create(DFSClient.java:1382)
2021-05-04T05:12:27.0111603Z 	at org.apache.hadoop.hdfs.DFSClient.create(DFSClient.java:1307)
2021-05-04T05:12:27.0112839Z 	at org.apache.hadoop.hdfs.DistributedFileSystem$6.doCall(DistributedFileSystem.java:384)
2021-05-04T05:12:27.0114327Z 	at org.apache.hadoop.hdfs.DistributedFileSystem$6.doCall(DistributedFileSystem.java:380)
2021-05-04T05:12:27.0115817Z 	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
2021-05-04T05:12:27.0117418Z 	at org.apache.hadoop.hdfs.DistributedFileSystem.create(DistributedFileSystem.java:380)
2021-05-04T05:12:27.0119023Z 	at org.apache.hadoop.hdfs.DistributedFileSystem.create(DistributedFileSystem.java:324)
2021-05-04T05:12:27.0120283Z 	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:908)
2021-05-04T05:12:27.0121363Z 	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:889)
2021-05-04T05:12:27.0122322Z 	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:786)
2021-05-04T05:12:27.0123271Z 	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:775)
2021-05-04T05:12:27.0124376Z 	at org.apache.hadoop.hbase.util.FSUtils.setVersion(FSUtils.java:649)
2021-05-04T05:12:27.0125570Z 	at org.apache.hadoop.hbase.util.FSUtils.setVersion(FSUtils.java:612)
2021-05-04T05:12:27.0127076Z 	at org.apache.hadoop.hbase.HBaseTestingUtility.createRootDir(HBaseTestingUtility.java:1028)
2021-05-04T05:12:27.0129077Z 	at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniHBaseCluster(HBaseTestingUtility.java:886)
2021-05-04T05:12:27.0132269Z 	at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniCluster(HBaseTestingUtility.java:862)
2021-05-04T05:12:27.0134533Z 	at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniCluster(HBaseTestingUtility.java:779)
2021-05-04T05:12:27.0136417Z 	at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniCluster(HBaseTestingUtility.java:750)
2021-05-04T05:12:27.0138346Z 	at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniCluster(HBaseTestingUtility.java:719)
2021-05-04T05:12:27.0139722Z 	at io.cdap.plugin.HBaseTest.beforeTest(HBaseTest.java:127)
2021-05-04T05:12:27.0140213Z 
2021-05-04T05:12:27.0141087Z testHBaseSink(io.cdap.plugin.HBaseTest)  Time elapsed: 0.68 sec  <<< ERROR!
2021-05-04T05:12:27.0142107Z java.lang.NoSuchMethodError: org.apache.hadoop.fs.FSOutputSummer.<init>(Ljava/util/zip/Checksum;II)V
2021-05-04T05:12:27.0143327Z 	at org.apache.hadoop.hdfs.DFSOutputStream.<init>(DFSOutputStream.java:1342)
2021-05-04T05:12:27.0144448Z 	at org.apache.hadoop.hdfs.DFSOutputStream.<init>(DFSOutputStream.java:1371)
2021-05-04T05:12:27.0145959Z 	at org.apache.hadoop.hdfs.DFSOutputStream.newStreamForCreate(DFSOutputStream.java:1403)
2021-05-04T05:12:27.0147289Z 	at org.apache.hadoop.hdfs.DFSClient.create(DFSClient.java:1382)
2021-05-04T05:12:27.0148274Z 	at org.apache.hadoop.hdfs.DFSClient.create(DFSClient.java:1307)
2021-05-04T05:12:27.0149465Z 	at org.apache.hadoop.hdfs.DistributedFileSystem$6.doCall(DistributedFileSystem.java:384)
2021-05-04T05:12:27.0151007Z 	at org.apache.hadoop.hdfs.DistributedFileSystem$6.doCall(DistributedFileSystem.java:380)
2021-05-04T05:12:27.0152941Z 	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
2021-05-04T05:12:27.0155985Z 	at org.apache.hadoop.hdfs.DistributedFileSystem.create(DistributedFileSystem.java:380)
2021-05-04T05:12:27.0158392Z 	at org.apache.hadoop.hdfs.DistributedFileSystem.create(DistributedFileSystem.java:324)
2021-05-04T05:12:27.0161043Z 	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:908)
2021-05-04T05:12:27.0162838Z 	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:889)
2021-05-04T05:12:27.0163905Z 	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:786)
2021-05-04T05:12:27.0165511Z 	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:775)
2021-05-04T05:12:27.0167493Z 	at org.apache.hadoop.hbase.util.FSUtils.setVersion(FSUtils.java:649)
2021-05-04T05:12:27.0169678Z 	at org.apache.hadoop.hbase.util.FSUtils.setVersion(FSUtils.java:612)
2021-05-04T05:12:27.0172625Z 	at org.apache.hadoop.hbase.HBaseTestingUtility.createRootDir(HBaseTestingUtility.java:1028)
2021-05-04T05:12:27.0176512Z 	at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniHBaseCluster(HBaseTestingUtility.java:886)
2021-05-04T05:12:27.0180461Z 	at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniCluster(HBaseTestingUtility.java:862)
2021-05-04T05:12:27.0184329Z 	at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniCluster(HBaseTestingUtility.java:779)
2021-05-04T05:12:27.0186625Z 	at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniCluster(HBaseTestingUtility.java:750)
2021-05-04T05:12:27.0189487Z 	at org.apache.hadoop.hbase.HBaseTestingUtility.startMiniCluster(HBaseTestingUtility.java:719)
2021-05-04T05:12:27.0191050Z 	at io.cdap.plugin.HBaseTest.beforeTest(HBaseTest.java:127)
```
